lib/srv/discovery/kube_integration_watcher.go

```diff
@@ -220,16 +226,48 @@ func (s *Server) getKubeAgentVersion(releaseChannels automaticupgrades.Channels)
 	return strings.TrimPrefix(agentVersion, "v"), nil
 }
 
-func (s *Server) getKubeIntegrationFetchers() []common.Fetcher {
+type IntegrationFetcher interface {
+	// GetIntegration returns the integration name that is used for getting credentials of the fetcher.
+	GetIntegration() string
+}
+
+func (s *Server) getKubeFetchers(integration bool) []common.Fetcher {
 	var kubeFetchers []common.Fetcher
 
-	s.muDynamicKubeIntegrationFetchers.RLock()
-	for _, fetcherSet := range s.dynamicKubeIntegrationFetchers {
-		kubeFetchers = append(kubeFetchers, fetcherSet...)
+	filter := func(fetcher common.Fetcher) bool {
+		f, ok := fetcher.(IntegrationFetcher)
+		if !ok {
+			// If fetcher doesn't provide integration field, return true if non-integration fetchers were requested.
+			return !integration
+		}
+		// If fetcher provides integration field, return true if integration fetchers were requested and integration
+		// field actually has a value.
+		return integration == (len(f.GetIntegration()) > 0)
+	}
```

Comment:
I'm not questioning the correctness of this function, but it is not clear at first sight that it is indeed doing the right thing.

I think creating two distinct functions would be easier to read
```go

	filterIntegrationFetchers := func(fetcher common.Fetcher) bool {
		f, ok := fetcher.(IntegrationFetcher)
		if !ok {
			return false
		}

		return f.GetIntegration() != ""
	}

	filterNonIntegrationFetchers := func(fetcher common.Fetcher) bool {
		f, ok := fetcher.(IntegrationFetcher)
		if !ok {
			return true
		}

		return f.GetIntegration() == ""
	}
```

Creating `GetIntegration` in the `common.Fetcher` interface would help as well.

Reply:
Refactored 👍 https://github.com/gravitational/teleport/pull/40155/commits/3eeed03f51159738b9501a62195c0a5a0c8682dd